### PR TITLE
Initial draft for a `cookiecutter` template (licenses and folder structure)

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
In specific, the MIT, Apache, and BSD-3-Clause licenses. I am adding them in a PR for visibility and for keeping track of changes more effectively; should we add templates for more licenses?

Edit: this PR also adds an initial structure for the template files to build upon further. The three licenses are adequate since we want to encourage the use of permissive licensing

### CODE REVIEW

**Review by Saransh-cpp — COMMENTED** [review]
I think it would also be nice to add cookiecutter.json and cookiecutter support through this PR, so that we can test this locally? (This looks like that it will work, but would be nice to have `cookiecutter.json` and `pyproject.toml` for the base repository, not the template.)

**Comment by agriyakhetarpal** [comment]
I did have an outline for a template ready to push. The template works and is a more bare-bones rendition of the `scientific-python` one (only `hatch` is supported at the moment).

I know this is now a larger PR for reviewing now (😅); I am happy to write separate, shorter PRs for further tasks such as adding utility functions in https://github.com/pybamm-team/pybamm-cookiecutter/issues/1#issuecomment-1660389638, adding entry points for parameter sets, and more

File: cookiecutter.json [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        "full_name": "Your name",
+        "email": "Your email",
+        "project_short_description": "A short description of your project",
+        "license": "Select a license",
+        "backend": {
+            "__prompt__": "Choose a build backend",
+            "hatch": "Hatchling (recommended for pure Python projects)"
+        }
+    }
+}
```
Comment (Saransh-cpp) [inline cookiecutter.json:43]:
```suggestion
}

```

File: docs/conf.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import importlib.metadata
+
+project = "pybamm-cookiecutter"
+copyright = "2023, Agriya Khetarpal"
+author = "Agriya Khetarpal"
```
Comment (Saransh-cpp) [inline docs/conf.py:5]: How about -

```py
project = "PyBaMM"
copyright = "2018-2023, The PyBaMM Team"
author = "The PyBaMM Team"
```

File: src/pybamm_cookiecutter/__init__.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+"""
+Copyright (c) 2023 Agriya Khetarpal. All rights reserved.
```
Comment (Saransh-cpp) [inline src/pybamm_cookiecutter/__init__.py:2]: Maybe we can attribute pybamm team here? Includes everyone!

File: src/pybamm_cookiecutter/__init__.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+pybamm-cookiecutter: A template for creating battery modeling projects based on PyBaMM
+"""
+
+
+__version__ = "0.1.0"
```
Comment (Saransh-cpp) [inline src/pybamm_cookiecutter/__init__.py:8]: Let's start with this

```suggestion
__version__ = "0.0.1"
```

File: {{cookiecutter.project_name}}/.pre-commit-config.yaml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+ci:
+  autoupdate_commit_msg: "chore: update pre-commit hooks"
+  autofix_commit_msg: "style: pre-commit fixes"
+
+repos:
+  - repo: https://github.com/psf/black
+    rev: "23.7.0"
+    hooks:
+      - id: black-jupyter
```
Comment (Saransh-cpp) [inline {{cookiecutter.project_name}}/.pre-commit-config.yaml:13]: We already have `ruff`!

```suggestion
```

File: {{cookiecutter.project_name}}/pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+  "Programming Language :: Python :: 3.8",
+  "Programming Language :: Python :: 3.9",
+  "Programming Language :: Python :: 3.10",
+  "Programming Language :: Python :: 3.11",  # PyBaMM supports >=3.8,<3.12
+  "Topic :: Scientific/Engineering",
+]
+{%- if cookiecutter.backend == "hatch" %}
+dynamic = ["version"]
+dependencies = ["pybamm"]
+{%- endif %}
```
Comment (Saransh-cpp) [inline {{cookiecutter.project_name}}/pyproject.toml:44]:
```suggestion
{%- endif %}
dependencies = ["pybamm"]
```

File: {{cookiecutter.project_name}}/pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+Homepage = "{{ cookiecutter.url }}"
+"Bug Tracker" = "{{ cookiecutter.url }}/issues"
+Discussions = "{{ cookiecutter.url }}/discussions"
+Changelog = "{{ cookiecutter.url }}/releases"
+
+[tool.hatch]
+version.path = "src/{{ cookiecutter.__project_slug }}/__init__.py"
+envs.default.dependencies = [
+  "pybamm",
+]
```
Comment (Saransh-cpp) [inline {{cookiecutter.project_name}}/pyproject.toml:68]:
```suggestion
{%- if cookiecutter.backend == "hatch" %}
[tool.hatch]
version.path = "src/{{ cookiecutter.__project_slug }}/__init__.py"
envs.default.dependencies = [
  "pybamm",
]
{%- endif %}
```

File: {{cookiecutter.project_name}}/src/{{cookiecutter.__project_slug}}/__init__.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+"""
+Copyright (c) {{ cookiecutter.__year }} {{ cookiecutter.full_name }}. All rights reserved.
+
+{{ cookiecutter.project_name }}: {{ cookiecutter.project_short_description }}
+"""
+
+
+__version__ = "0.1.0"
```
Comment (Saransh-cpp) [inline {{cookiecutter.project_name}}/src/{{cookiecutter.__project_slug}}/__init__.py:8]:
```suggestion
__version__ = "0.0.1"
```

File: src/pybamm_cookiecutter/__init__.py (hunk 4 above)
Comment (Saransh-cpp) [reply src/pybamm_cookiecutter/__init__.py:8]: till the time we switch to VCS versioning ;)

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
Awesome, thanks @agriyakhetarpal! This is the perfect sized PR, very minimalistic, but something that can be tested.

File: src/pybamm_cookiecutter/__init__.py (hunk 3 above)
Comment (agriyakhetarpal) [reply src/pybamm_cookiecutter/__init__.py:2]: Fair enough!

**Comment by agriyakhetarpal** [comment]
How should we test whether the cookiecutter works as intended? We could use a context manager along with [tempfile](https://docs.python.org/3/library/tempfile.html) to create cookiecutters with a specification and test whether `pyproject.toml` contains specific fields with `tomllib`, and the version of the created package could be tested upon importing it with `importlib.metadata`. Not sure how other cookiecutters do it

**Review by Saransh-cpp — APPROVED** [review]
Thanks @agriyakhetarpal@ Looks good! For the tests, we can skip them for now, but in the future we can use shell scripting to create the required package from `pybamm-cookiecutter` and check if it is installable. Similarly for licenses, we can check if the created package has the selected license.

**Pull request merged**